lms/static/scripts/frontend_apps/services/client-rpc.js

```diff
@@ -24,17 +59,37 @@ export class ClientRpc {
    * @param {Object} options
    *   @param {string[]} options.allowedOrigins -
    *     Origins that are allowed to request client configuration
-   *   @param {Object} options.clientConfig -
+   *   @param {string} options.authToken -
+   *     Auth token used in LMS backend API calls to refresh the grant token
+   *     in the `clientConfig` if it has expired
+   *   @param {ClientConfig} options.clientConfig -
    *     Configuration for the Hypothesis client. Whatever is provided here is
    *     passed directly to the client via `window.postMessage` when it requests
    *     configuration. It should be a subset of the config options specified at
    *     https://h.readthedocs.io/projects/client/en/latest/publishers/config/.
    */
-  constructor({ allowedOrigins, clientConfig }) {
+  constructor({ allowedOrigins, authToken, clientConfig }) {
     this._server = new Server(allowedOrigins);
 
-    // Handle the initial request for configuration from the Hypothesis client.
-    this._server.register('requestConfig', () => clientConfig);
+    // Handle the requests for configuration from the Hypothesis client.
+    this._server.register('requestConfig', async () => {
+      // Refresh the grant token used by the client to access the H API, if
+      // it expired.
+      const grantToken = clientConfig.services[0].grantToken;
+      const now = Date.now();
```

Comment:
Comparing `Date.now()` against the JWT token's "exp" field means there is an assumption that the user's clock is roughly correct. The use of SSL limits how far adrift the user's clock can be and still have the service work, but I don't know how tight that time window is. If the user's clock is ahead of the server we might end up refreshing it when we don't need to - that's OK. If the user's clock is behind it could fail to be refreshed when it needs to be. One way to guard against this would be to just refresh it regardless if the last refresh time was > N minutes ago, according to the LMS frontend.

Reply:
I think it could be worthwhile to add a brief comment in the code here for future devs.

Reply:
It might involve more network round trips but you could also change it to _always_ get a fresh grant token from the backend whenever the client requests one.